File: LinkedList.py
"""
#1.
#import collections
#llist=collections.deque()

#2.
#from colllection import deque
#llist=deque()

llist.append("bob")
llist.append("tom")
llist.append("sam")
llist.append("lee")


#Stack LIFO -  last in first out
#Queue FIFO - first in first out

# llist.append()        add to right end of list
# llist.appendleft()    add to left end of list

# llist.pop()           remove right-most element
# llist.popleft()       remove left-most element

print(llist)
print(llist[0])
print(llist[-1])
"""

# ___________
#/ varun     \
#\           /
# -----------
#            ^__^
#            (oo)\_______
#            (__)\       )\/\
#                ||----- |
#                ||     ||

class Node:
    def __init__(self,data):
        self.data=data
        self.next=None

class Linkedlist:
    def __init__(self):
        self.head=None

    def __iter__(self):
        node=self.head
        while node is not None:
            yield node
            node=node.next

    def __repr__(self):
        rep=""
        for node in self:
            rep+=node.data+","
        rep+="null"
        return rep

    def __getitem__(self,index):
        node=self.head
        count=0
        while (node is not None) and count<index:
            node=node.next
            count+=1;
        return node.data

    def enqueue(self, data):
        if self.head is None:
            self.head=Node(data)
            return

        for current in self:
            pass
        current.next=Node(data)

    def dequeue(self):
        if self.head is None:
            return

        temp=self.head
        self.head=self.head.next
        temp=None

    def push(self,data):
        node=Node(data)
        node.next=self.head
        self.head=node


    def pop(self):
        self.dequeue()



    def removeLast(self):
        if self.head is None:
            return
        if self.head.next is None:
            self.head=None
            return

        node=self.head
        while node.next.next is not None:
            node=node.next
        node.next=None


    def print(self):
        for current in self:
            print(current.data)

print("Stack")
stack=Linkedlist()
stack.push("Bob")
stack.push("Tom")
stack.pop()
stack.push("Lee")
stack.push("Sam")
stack.pop()
print(stack)

print("\nQueue")
queue=Linkedlist()
queue.enqueue("Bob")
queue.enqueue("Tom")
queue.enqueue("Lee")
queue.enqueue("Sam")
queue.dequeue()
print(queue)

print("\nsubscript testing")
print(queue[0])
print(queue[1])
print(queue[2])

File: Linkedlist.java
public class Linkedlist{
  private class Node{
    int val;
    Node next;

    public Node(){}
    public Node(int val){
      this.val=val;
    }
  }

  Node head;

  public Linkedlist(){}
  public Linkedlist(int val){
    this.head=new Node(val);
  }

  public Node prepend(int val){
    if (head==null){
      head=new Node(val);
      return head;
    }

    Node newNode=new Node(val);
    newNode.next=head;

    head=newNode;
    return head;
  }
  public Node append(int val){
    if (head==null){
      head=new Node(val);
      return head;
    }

    Node curr=head;
    while (curr.next!=null){
      curr=curr.next;
    }

    curr.next = new Node(val);
    return curr.next;
  }
  public void pop(){
    if (head==null){
      return;
    }
    head=head.next;
  }
  public void rpop(){
    if (head==null){
      return;
    }
    Node curr=head;
    while (curr.next.next!=null){
      curr=curr.next;
    }

    curr.next=null;
  }

  public void insert(Node prevNode){
    if (prevNode==null){
      return;
    }
    
  }

  public int size(){
    Node curr=head;
    int count=0;
    while (curr!=null){
      count++;
      curr=curr.next;
    }
    return count;
  }

  /*public Node insert(Node prevNode, int val){
      if (prevNode==null){
          prevNode=new Node(val);
          return prevNode;
      }

      Node temp=node;

      Node sentinel=new Node();
      sentinel.next=head;

      Node prev = sentinel;
      Node curr = head;

      for (int i=0;i<index;i++){
          prev=prev.next;
          curr=curr.next;
      }

  }*/

  public void push(int val){
    Node current=head;
    while (current!=null){
      current=current.next;
    }
    current.val=val;
    current.next=new Node();
  }

  public String toString(){
    String llist="";
    Node curr=head;
    while (curr!=null){
      llist+=curr.val+" ";
      curr=curr.next;
    }
    return llist;
  }
  public static void main(String[] args){
    Linkedlist llist= new Linkedlist();
    llist.append(5);
    llist.append(4);
    llist.prepend(3);
    llist.pop();
    llist.append(7);
    
    System.out.println(llist);
    System.out.println(llist.size());

  }
}
